### DISCUSSION TITLE
windows build

### DISCUSSION THREAD

**Discussion by GOBIKRISHNAN-R** [baresip/baresip#2004] [Q&A]
When i'm trying to build the baresip v2.0.2 release, I'm able to use the [baresip.vcxproj](https://github.com/baresip/baresip/blob/v2.0.2/mk/win32/baresip.vcxproj) for building the baresip application on visual studio. But there's no [baresip.vcxproj](https://github.com/baresip/baresip/blob/v2.0.2/mk/win32/baresip.vcxproj) file for building the baresip. 

Is it deleted after the 2.0.2 release?

If yes, can you please share me the build procedure for windows?

Thanks,
Gobi

**Comment by alfredh** [comment]
you have to use CMake

can you try this:

```
cmake .
cmake --build .
```

**Comment by chschnell** [comment]
Hi,

I tried building libre and baresip using cmake with MSYS2/Mingw64 and Visual 2022, in both cases neither TCP nor TLS work (UDP does). I've also build under Linux to cross-check, there everything works as expected. I've used REGISTER and INVITE requests in my tests with all three transports.

The error output I'm getting is (build with Visual 2022):

```shell
$ ./build/baresip.exe -4 -f ../.baresip
baresip v3.6.0 Copyright (C) 2010 - 2023 Alfred E. Heggestad et al.
Local network addresses:
 <GUID>:  <IP>
 <GUID>:  <IP>
aucodec: PCMU/8000/1
aucodec: PCMA/8000/1
aufilt: auconv
aufilt: auresamp
winwave: output devices: 3, input devices: 0
ausrc: winwave
auplay: winwave
vidsrc: dshow
dll: open: ./dshow.dll LoadLibraryA() failed
module dshow.dll: No such file or directory [2]
medianat: stun
medianat: turn
medianat: ice
Populated 1 account
Populated 3 contacts
Populated 2 audio codecs
Populated 2 audio filters
Populated 0 video codecs
Populated 0 video filters
baresip is ready.
reg: sip:<URI> (prio 0): Register: Invalid argument [22]
```

Can somebody please cross-check TCP support under Windows to confirm or disprove?

And, is Mingw64 officially supported by baresip (apart from one error in udp/udp.c which is easy to fix it builds fine and all self-tests pass)? I'm asking because libre from creytiv does.

Thanks for any help on this.

**Reply by sreimers** [reply]
MinGW-w64 is supported and builds fine with our CI:

https://github.com/baresip/re/actions/runs/6742949704/job/18329942776

Which build error do you get?

For TLS support, OpenSSL is required.

**Reply by chschnell** [reply]
Thank you for your reply.

It's not a build error, libre and baresip build fine and pass all self-tests, it's a runtime error (see the last line of the output of my earlier post).

**Comment by chschnell** [comment]
> MinGW-w64 is supported and builds fine with our CI:
> 
> https://github.com/baresip/re/actions/runs/6742949704/job/18329942776

As this is an Ubuntu/MinGW-w64 cross-build for Windows I decided to try that, too.

So I cross-build openssl, zlib, libre and baresip on a Debian 12 "bookworm" host using the MinGW-w64 toolchain, see attached `baresip-crossbuild.txt` for the exact sequence of commands I used and a few comments, I tried to keep it reproducable in case someone wants to retrace my steps.

I then copied the binaries and libraries to my Windows 10 PC and ran 3 tests from PowserShell:

1. baresip's selftest.exe, then 
1. baresip.exe with an account having `;transport=udp`, and then
1. baresip.exe with the same account but having `;transport=tcp`.

Again, the third test fails, see attached `baresip-run.txt` for the test's output and a few comments.

To summarize:

- I've now built libre+baresip for Windows in three different ways:
  - native on Windows using Visual 2022,
  - native on Windows using MSYS2/Mingw64 and
  - cross on Debian using MinGW-w64

  and in all three cases I observe the exact same pattern: Everything builds fine, all run-time tests pass, baresip.exe runs fine with an account using UDP, but fails when using TCP.
- On the other hand, if I build libre+baresip native under Linux it runs just fine (tested with Debian and Raspbian), meaning UDP, TCP and TLS work perfectly (I test against my own Kamailio server).

I've come to think that the problem sits inside libre (which is probably obvious anyway). I say that because I've been working on a little test using only libre from baresip, and there is where I first stumbled upon this: TCP/TLS work under Linux but fail under Windows. Once noticing that I began trying this with bearsip.exe, observed the same, and then I came here.
[baresip-crossbuild.txt](https://github.com/baresip/baresip/files/13257393/baresip-crossbuild.txt)
[baresip-run.txt](https://github.com/baresip/baresip/files/13257394/baresip-run.txt)

**Reply by sreimers** [reply]
I can reproduce this now. Looks like a race condition.

**Reply by chschnell** [reply]
Thank you!

While you're at it, there is a single minor problem in re/src/udp/udp.c when building under MSYS2/Mingw64 I'd like to report.

In line 32 it reads:

```C
typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
```

Under 64 bit, this typedef should be `ULONG` instead of `UINT32` which causes an error, my current work-around is to replace it with:

```C
#ifndef _WIN64
typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
#endif /* _WIN64 */
```

I can drop this line entirely under MSYS2/Mingw64 since both `QOS_FLOWID` and `PQOS_FLOWID` are properly defined elsewhere in the system headers.

Now I don't know about any side-effects regarding this change when building with Visual or when building cross from Linux as I didn't need to make these modifications there, under Visual and Linux-cross it just worked out of the box.

I'm not sure what to make of this, maybe you have an idea. If you want I can try this out under Visual and Linux-cross to check for any side-effects.

Other than that it libre and baresip build perfectly fine under MSYS2/Mingw64.

**Reply by sreimers** [reply]
Can you try if this works for MSYS2?

```diff
diff --git a/src/udp/udp.c b/src/udp/udp.c
index 496e7581..d9864567 100644
--- a/src/udp/udp.c
+++ b/src/udp/udp.c
@@ -28,12 +28,12 @@
 #include <re_sa.h>
 #include <re_udp.h>
 #ifdef WIN32
-#if !defined(_MSC_VER)
+#ifdef(__MINGW32__)
 typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
 #ifndef QOS_NON_ADAPTIVE_FLOW
 #define QOS_NON_ADAPTIVE_FLOW 0x00000002
 #endif
-#endif /*!_MSC_VER*/
+#endif /*__MINGW32__*/
 #include <qos2.h>
 #endif /*WIN32*/
```

**Reply by chschnell** [reply]
Thank you so much, I appreciate this a lot!

There's a small syntax issue:

```shell
$ cmake.exe --build build -j
[136/236] Building C object CMakeFiles/re-objs.dir/src/udp/udp.c.obj
FAILED: CMakeFiles/re-objs.dir/src/udp/udp.c.obj
D:\msys64\mingw64\bin\cc.exe -DARCH=\"AMD64\" -DHAVE_ATOMIC -DHAVE_GETOPT -DHAVE_INET6 -DHAVE_PTHREAD -DHAVE_SELECT -DHAVE_UNISTD_H -DHAVE_UNIXSOCK=1 -DOS=\"Windows\" -DRE_VERSION=\"3.6.1\" -DUSE_DTLS -DUSE_OPENSSL -DUSE_OPENSSL_AES -DUSE_OPENSSL_HMAC -DUSE_OPENSSL_SRTP -DUSE_TLS -DUSE_ZLIB -DVER_MAJOR=3 -DVER_MINOR=6 -DVER_PATCH=1 -DWIN32 -D_GNU_SOURCE -D_WIN32_WINNT=0x0600 -ID:/msys64/home/Christian/re.baresip/include -g -pedantic -Wall -Wbad-function-cast -Wcast-align -Wextra -Wmissing-declarations -Wmissing -prototypes -Wnested-externs -Wno-strict-aliasing -Wold-style-definition -Wshadow -Waggregate-return -Wstrict-prototypes -Wuninitialized -Wvla -MD -MT CMakeFiles/re-objs.dir/src/udp/udp.c.obj -MF CMakeFiles\re-objs.dir\src\udp\udp.c.obj.d -o CMakeFiles/re-objs.dir/src/udp/udp.c.obj -c D:/msys64/home/Christian/re.baresip/src/udp/udp.c
D:/msys64/home/Christian/re.baresip/src/udp/udp.c:31:7: error: macro names must be identifiers
   31 | #ifdef(__MINGW32__)
      |       ^
[153/236] Building C object CMakeFiles/re-objs.dir/src/pcp/msg.c.obj
ninja: build stopped: subcommand failed.
```

So I removed the braces `(` and `)` and replaced that line with:

```C
#ifdef __MINGW32__
```

But then it's the same logical error again as before, `__MINGW32__` is defined and the 32-Bit typedef collides with the 64-Bit typedef previously defined in system header `/msys64/mingw64/include/qos2.h`:

```shell
$ cmake.exe --build build -j
[136/236] Building C object CMakeFiles/re-objs.dir/src/udp/udp.c.obj
FAILED: CMakeFiles/re-objs.dir/src/udp/udp.c.obj
D:\msys64\mingw64\bin\cc.exe -DARCH=\"AMD64\" -DHAVE_ATOMIC -DHAVE_GETOPT -DHAVE_INET6 -DHAVE_PTHREAD -DHAVE_SELECT -DHAVE_UNISTD_H -DHAVE_UNIXSOCK=1 -DOS=\"Windows\" -DRE_VERSION=\"3.6.1\" -DUSE_DTLS -DUSE_OPENSSL -DUSE_OPENSSL_AES -DUSE_OPENSSL_HMAC -DUSE_OPENSSL_SRTP -DUSE_TLS -DUSE_ZLIB -DVER_MAJOR=3 -DVER_MINOR=6 -DVER_PATCH=1 -DWIN32 -D_GNU_SOURCE -D_WIN32_WINNT=0x0600 -ID:/msys64/home/Christian/re.baresip/include -g -pedantic -Wall -Wbad-function-cast -Wcast-align -Wextra -Wmissing-declarations -Wmissing-prototypes -Wnested-externs -Wno-strict-aliasing -Wold-style-definition -Wshadow -Waggregate-return -Wstrict-prototypes -Wuninitialized -Wvla -MD -MT CMakeFiles/re-objs.dir/src/udp/udp.c.obj -MF CMakeFiles\re-objs.dir\src\udp\udp.c.obj.d -o CMakeFiles/re-objs.dir/src/udp/udp.c.obj -c D:/msys64/home/Christian/re.baresip/src/udp/udp.c
In file included from D:/msys64/home/Christian/re.baresip/src/udp/udp.c:37:
D:/msys64/mingw64/include/qos2.h:17:15: error: conflicting types for 'QOS_FLOWID'; have 'ULONG' {aka 'long unsigned int'}
   17 | typedef ULONG QOS_FLOWID, *PQOS_FLOWID;
      |               ^~~~~~~~~~
D:/msys64/home/Christian/re.baresip/src/udp/udp.c:32:16: note: previous declaration of 'QOS_FLOWID' with type 'QOS_FLOWID' {aka 'unsigned int'}
   32 | typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
      |                ^~~~~~~~~~
D:/msys64/mingw64/include/qos2.h:17:28: error: conflicting types for 'PQOS_FLOWID'; have 'ULONG *' {aka 'long unsigned int *'}
   17 | typedef ULONG QOS_FLOWID, *PQOS_FLOWID;
      |                            ^~~~~~~~~~~
D:/msys64/home/Christian/re.baresip/src/udp/udp.c:32:29: note: previous declaration of 'PQOS_FLOWID' with type 'PQOS_FLOWID' {aka 'unsigned int *'}
   32 | typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
      |                             ^~~~~~~~~~~
[153/236] Building C object CMakeFiles/re-objs.dir/src/pcp/option.c.obj
ninja: build stopped: subcommand failed.
```

This patch works for me (it disables these defines on a 64-Bit installation like mine):

```patch
diff --git a/src/udp/udp.c b/src/udp/udp.c
index 496e758..912171a 100644
--- a/src/udp/udp.c
+++ b/src/udp/udp.c
@@ -28,12 +28,12 @@
 #include <re_sa.h>
 #include <re_udp.h>
 #ifdef WIN32
-#if !defined(_MSC_VER)
+#if defined(__MINGW32__) && !defined(__MINGW64__)
 typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
 #ifndef QOS_NON_ADAPTIVE_FLOW
 #define QOS_NON_ADAPTIVE_FLOW 0x00000002
 #endif
-#endif /*!_MSC_VER*/
+#endif /*__MINGW32__*/
 #include <qos2.h>
 #endif /*WIN32*/
```

**Reply by sreimers** [reply]
Ok you can try now these PRs:

https://github.com/baresip/re/pull/1002
https://github.com/baresip/re/pull/1001

**Reply by chschnell** [reply]
I can confirm that these PRs fix both issues.

I can now build under Windows (MSYS2/Mingw64) without needing to change udp.c, and baresip.exe built under Windows (MSYS2/Mingw64) now correctly supports all three protocols UDP, TCP and TLS.

Great!
